lessons/00-intro/best-practices/best_practices_guide.md

```diff
@@ -0,0 +1,260 @@
+
+### Pre-commit Hooks
+
+Git provides a useful tool to help putting these technique into practice. Pre-commit hooks allow you to perform these code checks every time you make a commit. The `pre-commit` packages helps to use this tool, by creating a customizable config file for the pre-commits. You can find it [here](./.pre-commit-config.yaml).
+
+Before using the pre-commits, you must install them, by running in your terminal:
+
+```bash
+pre-commit install -t pre-commit
+```
+
+### Demo
+
+- Run `isort` to format the imports
+- Run `black` to format the code
+- Create a commit with the changes and see that all check pass
+
+## Git
+
+Working with a clean and legible git history is key to rendering your commit history usable. The whole point of using `git` is to keep track of the changes in code and collaborate better. Achieving that requires establishing rules that go beyond the simple practices we keep when working alone.
+
+* If you want to learn more about the branching strategies most commonly used today: [what are the Best Git Branching Strategies](https://www.abtasty.com/blog/git-branching-strategies/) goes through the most commonly used branching strategies
+* More details on the Gitflow framework: [Gitflow Workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow)
+* If you want to master more advanced git functionalities: [Learning Git Branching](https://learngitbranching.js.org/)
+
+## Continuous Integration
+
+## Continuous Integration (CI)
+
+Continuous Integration (CI) is a software development practice that involves frequently integrating code changes made by multiple developers into a shared repository. The main goal of CI is to streamline and automate the process of integrating code, running tests, and identifying and fixing bugs early in the development cycle. CI ensures that all changes are tested and incorporated into the main codebase in a consistent and reliable manner. It helps in maintaining code quality, reducing integration issues, and enhancing collaboration among team members. By automating the integration process, CI also enables faster feedback loops, making it easier to catch and resolve issues promptly, leading to improved software stability and quicker delivery of new features. Overall, CI is essential for achieving efficient, robust, and reliable software development.
+
+We included an example CI workflow in the [./.github/workflows/ci.yaml](./.github/workflows/ci.yaml) file. It does:
+
+* Copy the current repo into the GitHub Action
+* Set up python
+* Install requirements: ensures there are no conflicts in the requirements
+* Run pre-commit hooks: ensures developer installed pre-commit hooks and thus linted and formatted the code
+
+But a CI can also:
```

Comment:
I think you can insist on the distinction between CI, CD and github actions, which is how we implement these in github